miden-crypto/src/merkle/sparse_path.rs

```diff
@@ -0,0 +1,677 @@
+    /// # let zero = RpoDigest::new([ZERO; 4]);
+    /// # let one = RpoDigest::new([ONE; 4]);
+    /// # let sparse_path = SparseMerklePath::from_sized_iter(vec![zero, one, one, zero]).unwrap();
+    /// let depth = NonZero::new(sparse_path.depth()).unwrap();
+    /// assert_eq!(
+    ///     sparse_path.at_idx(0).unwrap(),
+    ///     sparse_path.at_depth(depth).unwrap(),
+    /// );
+    /// ```
+    pub fn at_idx(&self, index: usize) -> Option<RpoDigest> {
+        // If this overflows *or* if the depth is zero then the index was out of bounds.
+        let depth = NonZero::new(u8::checked_sub(self.depth(), index as u8)?)?;
+        self.at_depth(depth).ok()
+    }
+
+    // PROVIDERS
+    // ============================================================================================
+
+    /// Constructs a borrowing iterator over the nodes in this path.
+    pub fn iter(&self) -> impl ExactSizeIterator<Item = RpoDigest> {
+        self.into_iter()
+    }
+
+    // PRIVATE HELPERS
+    // ============================================================================================
+
+    /// Constructs a sparse Merkle path from a manually specified tree depth, and an iterator over
+    /// Merkle nodes from deepest to shallowest.
+    ///
+    /// Knowing the size is necessary to calculate the depth of the tree, which is needed to detect
+    /// which nodes are empty nodes.
+    ///
+    /// Warning: this method does not check if the iterator contained more elements.
+    ///
+    /// # Errors
+    /// Returns [MerkleError::DepthTooBig] if `tree_depth` is greater than [SMT_MAX_DEPTH].
+    fn from_iter_with_depth(
+        tree_depth: u8,
+        iter: impl IntoIterator<Item = RpoDigest>,
+    ) -> Result<Self, MerkleError> {
+        if tree_depth > SMT_MAX_DEPTH {
+            return Err(MerkleError::DepthTooBig(tree_depth as u64));
+        }
+
+        let mut empty_nodes_mask: u64 = 0;
+        let mut nodes: Vec<RpoDigest> = Default::default();
+
+        for (depth, node) in iter::zip(path_depth_iter(tree_depth), iter) {
+            let &equivalent_empty_node = EmptySubtreeRoots::entry(tree_depth, depth.get());
+            let is_empty = node == equivalent_empty_node;
+            let node = if is_empty { None } else { Some(node) };
+
+            match node {
+                Some(node) => nodes.push(node),
+                None => empty_nodes_mask |= Self::bitmask_for_depth(depth),
+            }
+        }
+
+        Ok(SparseMerklePath { nodes, empty_nodes_mask })
+    }
+
+    const fn bitmask_for_depth(node_depth: NonZero<u8>) -> u64 {
+        // - 1 because paths do not include the root.
+        1 << (node_depth.get() - 1)
+    }
+
+    const fn is_depth_empty(&self, node_depth: NonZero<u8>) -> bool {
+        (self.empty_nodes_mask & Self::bitmask_for_depth(node_depth)) != 0
+    }
+
+    fn get_nonempty_index(&self, node_depth: NonZero<u8>) -> usize {
```

Comment:
Let's add doc comments here to explain what this function does. My understanding is that it takes node depth and returns the node's index in the `self.nodes` vector - correct? And if the node at this depth is empty, would it panic?

An alternative could be to combine this method with `is_depth_empty()` to look something like this:
```rust
fn get_node_index(&self, node_depth: NonZero<u8>) -> Option<usize> {
    ...
}
```
If the node at the specified depth is empty, it would just return `None`. I think this could lead a slightly cleaner code elsewhere.

Reply:
Good catch! It wouldn't panic, it would return incorrect results. I've kept the name (I think it better represents the intent), but added a doc comment and changed the return type to `Option<size>`. It indeed makes the call site cleaner.